src/ucp/wireup/wireup_cm.c

```diff
@@ -257,9 +261,93 @@ static void ucp_cm_client_connect_cb(uct_ep_h ep, void *arg,
     }
 }
 
+/*
+ * Internal flush completion callback which is a part of close protocol,
+ * this flush was initiated by remote peer in disconnect callback on CM lane.
+ */
+static void ucp_ep_cm_disconnect_flushed_cb(ucp_request_t *req)
+{
+    ucp_ep_h ucp_ep = req->send.ep;
+
+    UCS_ASYNC_BLOCK(&ucp_ep->worker->async);
+    ucp_ep_cm_disconnect_cm_lane(ucp_ep);
+    ucp_request_complete_send(req, UCS_OK);
+    UCS_ASYNC_UNBLOCK(&ucp_ep->worker->async);
+}
+
+static unsigned ucp_ep_cm_do_disconnect(void *arg)
+{
+    ucp_ep_h ucp_ep     = (ucp_ep_h)arg;
+    ucp_worker_h worker = ucp_ep->worker;
+    void *req;
+
+    UCS_ASYNC_BLOCK(&worker->async);
+
+    ucs_trace_func("ucp_ep = %p, ucp_ep->flags = %xu", ucp_ep, ucp_ep->flags);
+
+    ucs_assert(ucp_ep_get_cm_uct_ep(ucp_ep) != NULL);
+
+    if (!(ucp_ep->flags & UCP_EP_FLAG_LOCAL_CONNECTED)) {
+        /* while we were in slow progress queue, the EP was flushed and
+         * disconnected from API call. Destroy the EP, complete close request.*/
+        ucs_assert(ucp_ep->flags & UCP_EP_FLAG_CLOSE_REQ_VALID);
+        ucp_ep_do_disconnect(ucp_ep_ext_gen(ucp_ep)->close_req.req);
+        goto out;
+    }
+
+    req = ucp_ep_flush_internal(ucp_ep, UCT_FLUSH_FLAG_LOCAL, NULL, 0, NULL,
```

Comment:
add TODO that we should set EP to error state?

Reply:
in which case?
if `ucp_ep_flush_internal` fails, EP will be set failed at line 311

Reply:
after getting disconnect we should set to err state
to prevent user from sending more opts after we already started to flush